src/utils/front-matter.js

```diff
@@ -1,55 +1,34 @@
 "use strict";
 
-const escape = require("escape-string-regexp");
 const {
   builders: { hardline, markAsRoot },
 } = require("../document");
 
-const YAML_DELIMITER = "---";
 // In some markdown processors such as pandoc,
 // "..." can be used as the end delimiter for YAML front-matter.
-const YAML_END_DELIMITER = "...";
-const DELIMITER_MAP = {
-  [YAML_DELIMITER]: "yaml",
-  "+++": "toml",
-};
 
-function parse(text) {
-  const startDelimiterRegex = Object.keys(DELIMITER_MAP).map(escape).join("|");
-  const endDelimiterRegex = [...Object.keys(DELIMITER_MAP), YAML_END_DELIMITER]
-    .map(escape)
-    .join("|");
-
-  const match = text.match(
-    // trailing spaces after delimiters are allowed
-    new RegExp(
-      `^(${startDelimiterRegex})([^\\n]*)\\n(?:|([\\s\\S]*?)\\n)(${endDelimiterRegex})[^\\n\\S]*(\\n|$)`
-    )
-  );
+// trailing spaces after delimiters are allowed
+const frontMatterRegex = /^(?<startDelimiter>-{3}|\+{3})(?<language>[^\n]*)\n(?:|(?<value>[\S\s]*?)\n)(?<endDelimiter>\k<startDelimiter>|\.{3})[^\S\n]*(?:\n|$)/;
```

Comment:
I'd rewrite it for readability:
```
const frontMatterRegex = new RegExp([
  "^(?<startDelimiter>-{3}|\\+{3})" +
    "(?<language>[^\\n]*)\\n" +
    "(?:|(?<value>[\\S\\s]*?)\\n)" +
    "(?<endDelimiter>k<startDelimiter>|\\.{3})" +
    "[^\\S\\n]*(?:\\n|$)",
]);
```

Reply:
I hate double escape regular expressions. Maybe I can use `Regex#source` to make it smore readable?

Reply:
Nobody likes that escaping, but IMHO it's the lesser evil.

(I wish JS had free-spacing regexps (`x` flag). Prettier would be able to reformat such regexps...)

> Maybe I can use Regex#source to make it more readable?

Can't imagine what you mean.

Reply:
```js
const frontMatterRegex = new RegExp(
  /^(?<startDelimiter>-{3}|\+{3})/.source +
    /(?<language>[^\n]*)/.source +
    /\n(?:|(?<value>[\S\s]*?)\n)/.source +
    "(?<endDelimiter>\\k<startDelimiter>|\\.{3})" +
    /[^\S\n]*(?:\n|$)/.source
);
```

Reply:
Put "\n" on 3rd part, prevent `unicorn/better-regex` remove `(?:` on that group. 4th part can't write in regex literal "invalid group".

Reply:
This lib looks cool http://xregexp.com/